python_transport/wirepas_gateway/protocol/mqtt_wrapper.py

```diff
@@ -288,15 +299,33 @@ def publish(self, topic, payload, qos=1, retain=False) -> None:
         Args:
             topic: Topic to publish on
             payload: Payload
-            qos: Qos to use
-            retain: Is it a retain message
+            qos: Qos to use. Can be less than requested if broker does
+                not support it
+            retain: Is it a retain message. Can be discarded if broker
+                    does not support it
 
         """
+        # No need to check qos or retain at this stage as
+        # done later in real publish to broker
+
         # Send it to the queue to be published from Mqtt thread
         self._publish_queue.put((topic, payload, qos, retain))
         self._publish_monitor.on_publish_request()
 
     def subscribe(self, topic, cb, qos=2) -> None:
```

Comment:
General comment around Qos1. Although that is probably something to handle in the callbacks rather than in this part of the code. Should we think of a mechanism to handle duplicates? I am thinking specifically about instructions coming in that would result in WM duplicate instructions. For example, it might be possible to receive to sendData to WM stack.

Reply:
yes definitely something that could happen.